netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java

```diff
@@ -94,12 +113,50 @@ protected final void initChannel(Http2HeadersDecoder headersDecoder) throws Exce
 
     handler = newHandler();
 
-    channel = new EmbeddedChannel(handler);
+    channel = new FakeClockSupportedChanel(handler);
     ctx = channel.pipeline().context(handler);
 
     writeQueue = initWriteQueue();
   }
 
+  private final class FakeClockSupportedChanel extends EmbeddedChannel {
+    EventLoop eventLoop;
+
+    FakeClockSupportedChanel(ChannelHandler... handlers) {
+      super(handlers);
+    }
+
+    @Override
+    public EventLoop eventLoop() {
+      if (eventLoop == null) {
+        createEventLoop();
+      }
+      return eventLoop;
+    }
+
+    void createEventLoop() {
+      EventLoop realEventLoop = super.eventLoop();
+      if (realEventLoop == null) {
+        return;
+      }
+      eventLoop = mock(EventLoop.class, delegatesTo(realEventLoop));
+      ArgumentMatcher<Runnable> isFakeClockRunnable = new ArgumentMatcher<Runnable>() {
+        @Override
+        public boolean matches(Object argument) {
+          String name = argument.toString();
+          for (String taskName : scheduleTheseWithFakeClock) {
+            if (name.contains(taskName)) {
+              return true;
+            }
+          }
+          return false;
+        }
+      };
+      doAnswer(delegatesTo(fakeClock.getScheduledExecutorService())).when(eventLoop)
+          .schedule(argThat(isFakeClockRunnable), anyLong(), any(TimeUnit.class));
```

Comment:
Everything calling _schedule_ should be safe to delay, because it is based on time. It seems every usage of schedule should be able to use the fake clock, with the possible (unlikely) exception of when something is scheduled with 0 or negative delay (in the past).